### ISSUE TITLE
[BUG] Heart Rate drops to 0 every other second in Peloton Apple TV app

### ISSUE THREAD

**Issue by hoveeman** [cagnulein/qdomyos-zwift#636]
**Describe the bug**
While broadcast heart rate from QZ to the Peloton app, the heart rate will drop to 0 every other second. Video attached. This makes the statistics in Peloton be incorrect. Inside of QZ my heart rate is accurate the whole time and never drops to 0. This is a new bug as I've been using the same heart rate monitor [Scosche Rhythm 24](https://www.scosche.com/rhythm24-waterproof-armband-heart-rate-monitor) for about a year with no problems.

I tested my HR monitor directly connected to the Peloton app and it worked properly. I also tried force quitting the Peloton app and reconnecting to QZ. I did however notice a difference between connecting my QZ iPad app to the Peloton app. Peloton will show the name of the HR monitor (QZ app name) when it connects. There are 2 different names that will show up in Peloton from QZ

- `QZ`- Heart rate drops to 0 when it shows it connected to heart rate `QZ`. It does not show battery percentage of my iPad.
- `qDomyos-zwift` - Heart rate works correctly in Peloton app when it shows it is connected to heart rate `qDomyos-zwift`, It also shows battery percentage of my iPad. Picture attached of correct working Heart Rate name

Note I tried toggling on/off the heart rate options in settings and couldn't find any correlation. Items I turned on/off
- `Heart Rate service outside FTMS`
- `Disable HRM from Machinery`
- `Disable KCal from Machinery`

**To Reproduce**
- Connect HR Monitor to QZ
- Verify HR is showing in a Tile
- Go to Peloton app on a different device and connect HR
- Select device running QZ - In my case, `iPad Pro`
- If HR name shows as `QZ` - verify heart rate drops to 0 every other second in a ride
- If HR name shows as `qDomyos-zwift` - verify heart rate is correct for the whole ride and doesn't drop.

**Expected behavior**
Heart Rate doesn't drop to 0 in the Peloton app

**Screenshots**
Correct name showing up in Peloton app
![qDomyos-zwift Heart Rate Name](https://user-images.githubusercontent.com/18269595/151984360-398dc539-bb79-457e-acfa-1bd1e29514eb.jpg)

Video of the Heart Rate drops when heart rate name is `QZ`
https://user-images.githubusercontent.com/18269595/151984397-007988fd-7946-44a2-9dd4-609f1185ea0e.mp4

**Smartphone (please complete the following information):**
 - Device: iPad Pro 10.5
 - OS: iOS 15.2.1
 - Version 2.10.10

**Append a debug log**
Debug from ride that dropped heart rate to 0 every other second in Peloton App. Heart rate was perfect in QZ Tile
[QZ Heart Rate Name debug-Tue_Feb_1_06_14_49_2022.log](https://github.com/cagnulein/qdomyos-zwift/files/7979063/QZ.Heart.Rate.Name.debug-Tue_Feb_1_06_14_49_2022.log)

Not sure if this is useful, but this is a debug when I got Peloton to see the heart rate as `qDomyos-zwift` 
[qDomyos-zwift Heart Rate Name debug-Tue_Feb_1_07_01_28_2022.log](https://github.com/cagnulein/qdomyos-zwift/files/7979073/qDomyos-zwift.Heart.Rate.Name.debug-Tue_Feb_1_07_01_28_2022.log)

**Comment by cagnulein** [comment]
@hoveeman when did the issue start? I have an idea about this.

**Comment by hoveeman** [comment]
@cagnulein Today was the second time I've noticed it. That video I shared above was from the first time I saw it on January 27th 2022.

**Comment by cagnulein** [comment]
@hoveeman so it doesn't happen all the time?

**Comment by hoveeman** [comment]
@cagnulein That's correct... It does not happen all the time. Last weekend when we were testing the Peloton Run issues it did not occur.

This morning I thought maybe it'd be related to Bike or Run. I tested a run this morning and it dropped to 0 as well. So no correlation between Bike or Run.

**Comment by cagnulein** [comment]
@hoveeman did you try to reboot your peloton device before doing a session?

**Comment by hoveeman** [comment]
@cagnulein Yes, I tried that and it didn't make a difference. I'm not sure why Peloton would see the Heart Rate info as `QZ`sometimes and `qDomyos-zwift` other times. I couldn't find any pattern/correlation.

**Comment by cagnulein** [comment]
@hoveeman QZ for run and qdomyos for ride i guess, isn't it?

**Comment by cagnulein** [comment]
i guess it's a cache issue on the ipad, but i don't think it's releated...or maybe yes? :) in peloton did you pair it everytime or does it automatically?

**Comment by hoveeman** [comment]
@cagnulein The first time, it auto connected. Then the next 5 times I would manually connect it, and also try to disconnect it and re-connect it.

**Comment by cagnulein** [comment]
@janabouc did you experience a similar issue? i would like to understand if it's a spread issue

**Comment by janabouc** [comment]
I don't connect my HRM this way. My Polar H10 can connect to 2 Bluetooth devices so it connects to both QZ and to Peloton separately. 

I use the ATV Peloton app, I don't stream from iPad to ATV while using QZ on my iPhone. I've never seen my battery level display on Peloton. I suggest OP try using Apple TV app instead of AirPlay (if I understand their setup). 

It might be hard for me to test this because Peloton automatically connects to my Polar and doesn't show other options like QZ as HRM.

**Comment by cagnulein** [comment]
@hoveeman my idea behind this is to try to bisect if it's a QZ issue or a peloton issue :)

**Comment by hoveeman** [comment]
@cagnulein For sure! I will test and report back with logs.

**Comment by janabouc** [comment]
> ok no problem @janabouc 
> 
> @hoveeman i saw now that you're using it over Apple TV. Did you try on an ipad too?

On my last treadmill run I used Peloton app on my iPad Pro with QZ floated over it, both on my iPad, no ATV. Heart rate was fine, I could see it on both, with no fluctuation.

**Comment by cagnulein** [comment]
Thanks @hoveeman great report!did you restart peloton app between workouts?

**Comment by hoveeman** [comment]
@cagnulein No, I did not. The Peloton app on my iPad was open the whole time. I did stop and start classes though.

**Comment by cagnulein** [comment]
So that's why. Passing from treadmill to ride it changes the Bluetooth fingerprint on qz but the peloton app can't know it. I guess a restart should fix it

**Comment by hoveeman** [comment]
Hmm. This morning on my Apple TV, I force quit peloton and relaunched it and still had the same issue. 

While on the iPad, it went to 0 after doing the treadmill, and I kept peloton open and then I force quit QZ on my phone 2 times after connecting to my bike and then it worked.

**Comment by cagnulein** [comment]
Ok noted. I will try to put Qz on both profiles

**Comment by hoveeman** [comment]
Saw the stalebot got ahold of this issue. I am still experiencing this issue and had it occur yesterday.

**Comment by cagnulein** [comment]
Pinned, thanks @hoveeman 
I'm trying to do my best in these days but it's really hard. Too much tickets every day including Facebook, mail, Reddit..

**Comment by cagnulein** [comment]
@hoveeman i did a modification about this today. Now with the peloton cadence sensor setting OFF you should be able to run Zwift and Peloton without changing everytime it. And also i guess it should fix the peloton issue (if i understood well the issue). Send me an email as a reminder for it, thanks

**Comment by hoveeman** [comment]
@cagnulein Apologies if I came across as "why isn't this done?!" I just wanted to bump the thread so it didn't get closed. Really appreciate you taking a look at it! I had a chance to try out the new build this morning. I disabled peloton cadence sensor. Heart rate worked properly sharing to the apple tv. I did a cycle, tread, back to cycle, and the heart rate worked every time. 

However, I did experience a new issue today with following peloton resistance for a class. I will open a new ticket related to that. With my testing today, I believe you fixed the HR issue and will close this for now. Thank you as always!
